### PR TITLE
Add support for documentation URL previews

### PR DESCRIPTION
**Pull request by Cadair**
I am not going to have time to fix this up, so please push any changes to this branch as you desire.

### CODE REVIEW

File: CHECKS
Comment (Cadair) [inline CHECKS:1]: This is useful when deploying using Dokku, so it would be a good addition to the repo for me.

File: .circleci/config.yml
Comment (bsipocz) [inline .circleci/config.yml:8]: we need to change this

File: .circleci/config.yml
Comment (bsipocz) [inline .circleci/config.yml:20]: also change this

File: .circleci/config.yml
Comment (Cadair) [reply .circleci/config.yml:8]: technically you *don't* as it's just a check to see if the artifact works, it's just any html page you like that it then reports as a status to this repo. If you don't want to test this blueprint on this repo then you can delete the whole file.

File: changebot/blueprints/circleci.py
Comment (bsipocz) [inline changebot/blueprints/circleci.py:71]: what does `*` do here? you don't seem to use any other parameter but the listed ones.

Comment (Cadair) [reply changebot/blueprints/circleci.py:71]: it makes `headers` a mandatory keyword argument.

**Review by bsipocz — CHANGES_REQUESTED** [review]
We need to add some testing for this if possible before letting it loose on astropy.

File: changebot/blueprints/circleci.py
Comment (pllim) [reply changebot/blueprints/circleci.py:71]: Why not just make it a regular position argument, then? (Just curious)

**Comment by pllim** [comment]
I second @bsipocz 's comment on adding tests.

File: changebot/blueprints/circleci.py
Comment (Cadair) [reply changebot/blueprints/circleci.py:71]: I can't really remember, I think I was copying the `headers=` syntax from requests, but it needed to be provided. :man_shrugging:

File: .circleci/config.yml
Comment (pllim) [reply .circleci/config.yml:8]: This assumes that `http://sunpy.org/index.html` is going to be up indefinitely, which is out of the bot's control. Might be better to get https://astropy.stsci.edu/data/index.html ?

File: .circleci/config.yml
Comment (pllim) [reply .circleci/config.yml:20]: I agree that this URL has to be changed but I don't know what to. Is this something that needs to be set up for the bot?

File: CHECKS
Comment (pllim) [reply CHECKS:1]: Useful in what way? Is it possible to document the reason for this file in the file itself for future maintainer(s)?

File: changebot/blueprints/circleci.py
Comment (pllim) [inline changebot/blueprints/circleci.py]: Typo: `chagebot` -> `changebot`

File: changebot/blueprints/circleci.py
Comment (pllim) [inline changebot/blueprints/circleci.py:10]: This is already defined in `changebot.github.github_api.HOST`.

File: changebot/blueprints/circleci.py
Comment (pllim) [reply changebot/blueprints/circleci.py:71]: Can we just make it a positional argument to make life simpler, then? :grimacing:

File: changebot/github/github_auth.py
Comment (pllim) [inline changebot/github/github_auth.py:118]: I am not sure what this is doing. So I'll defer this portion to @astrofrog .

**Review by pllim — COMMENTED** [review]
Overall, you might need to add in some `yield` like #55 to avoid timeout.

I am thinking we might be able to test some of the individual functions (see list below) using mock. I already caught an import typo by eye, so I still think it is important to add tests despite painfulness of mocking network response.

* `get_artifacts_from_build()`
* `get_documentation_url_from_artifacts()`
* `set_commit_status()`
* `set_status()`
* `get_installation_id_for_repo()`
* `generate_repo_id_map()`

As for the `webapp` part, I don't think we can test it until it is live. But the worst case is people get broken link to artifact (at least I hope nothing worse will happen...), so this risk is probably acceptable provided the unit tests above passed.

@astrofrog , do you have anything to add?

**Comment by pllim** [comment]
p.s. The import typo is causing CI failures at https://travis-ci.org/astropy/astropy-bot/jobs/376456518

File: .circleci/config.yml
Comment (bsipocz) [reply .circleci/config.yml:8]: or even better http://www.astropy.org/index.html

**Comment by bsipocz** [comment]
@pllim - Thanks for the review, I'll address the points. Also will rename `circleci` --> `circleci_docs_built` to better describe what it is for.

**Comment by Cadair** [comment]
@bsipocz I named it `circleci` because the root of it is a generic circle ci webhook handler. Just because the only thing it currently does with that web hook it extract the documentation artifact dosen't mean that's all it's good for :wink: up to you though.

File: .circleci/config.yml
Comment (Cadair) [reply .circleci/config.yml:8]: *takes offence and leaves* :wink:

File: CHECKS
Comment (Cadair) [reply CHECKS:1]: actually this should have something in it... *looks around for his file*

File: .circleci/config.yml
Comment (Cadair) [reply .circleci/config.yml:20]: it has to be the URL of the circleci blueprint. The blueprint listens for circleci webhooks which are sent to the address given in the config. (this URL is my running deployment of Giles)

File: .circleci/config.yml
Comment (pllim) [reply .circleci/config.yml:8]: Offence unintended! But just that the bot is more closely associated with astropy core currently. I like the sun!

**Comment by bsipocz** [comment]
@Cadair - ok, fair enough. Though then call it circleci_handler, or basically anything but purely circleci?

**Comment by Cadair** [comment]
Seems reasonable. `circleci_webhook_handler` makes the most sense to me.

File: changebot/blueprints/circleci.py
Comment (astrofrog) [inline changebot/blueprints/circleci.py:73]: @Cadair - why not use https://github.com/astropy/astropy-bot/blob/master/changebot/github/github_api.py#L466?

File: changebot/github/github_auth.py
Comment (astrofrog) [reply changebot/github/github_auth.py:118]: Looks good but please add a docstring :)

**Review by astrofrog — CHANGES_REQUESTED** [review]
I'm in favor of this! A few comments:

* I think since it would be difficult to write formal tests, just having CircleCI enabled and making sure this works in PRs would be fine (but unfortunately the code here won't run until this is merged!)

* I think the blueprint should be ``docpreview_circleci`` since we should make it clear it's for documentation previews. We might have other CircleCI-related blueprints in future.

* How do users opt-in/out of this? A section is needed in the README.

File: changebot/blueprints/circleci.py
Comment (Cadair) [reply changebot/blueprints/circleci.py:73]: This and that should probably be refactored so the code isn't duped. The reason I didn't use that method originally is that I am setting status on a *commit* not on a PR, so I didn't want to / couldn't create a PR object.

**Comment by Cadair** [comment]
@astrofrog 

> We might have other CircleCI-related blueprints in future.

As I mentioned to @bsipocz in another thread in theory this blueprint should be the home to all the circle ci webhook handlers so I think it's best to keep it named as such.

**Comment by astrofrog** [comment]
> As I mentioned to @bsipocz in another thread in theory this blueprint should be the home to all the circle ci webhook handlers so I think it's best to keep it named as such.

Ok sounds good.

So I guess one question is, is it useful to have astropy-bot active in the astropy-bot repository? Maybe the answer is 'no' since changes in PRs will never be reflected until the PR is merged?

**Comment by Cadair** [comment]
> How do users opt-in/out of this? A section is needed in the README.

Agreed, opting in is easy, you have to explicitly put the URL of the blueprint in your `.circleci` configs.

File: changebot/blueprints/circleci.py
Comment (bsipocz) [reply changebot/blueprints/circleci.py:73]: While addressing @pllim's review, I've actually started that refactoring last night :)

**Comment by pllim** [comment]
Re: tests -- If they turn out to be too complicated, let's do as Tom R suggested in his review and open a follow-up issue for testing instead. 

:+1: on adding documentation.

**Comment by astrofrog** [comment]
I think the plan is now to just keep the documentation build stuff in giles, so it doesn't make sense to keep this PR open (no point in maintaining this code in both giles and here). Some of this will move to baldrick too.

**Pull request closed without merging**
